Dev

DenisHusainov opened the pull request.

DenisHusainov committed:
Create BonusDoors and borders

DenisHusainov committed:
Fixed BonusDoors script

DenisHusainov committed:
delete log from script

dshestel commented on Assets/Scripts/BonusController.cs:
```diff
@@ -0,0 +1,23 @@
+
+public class BonusController : MonoBehaviour
+{
+    public static event Action Spawed = delegate { };
```
пиши без офографических ошибок

dshestel commented on Assets/Scripts/BonusController.cs:
```diff
@@ -0,0 +1,23 @@
+    public static event Action Spawed = delegate { };
+
+    [SerializeField]
+    private TextMeshPro textMeshPro = null;
```
1. не по код стайлу
2. название переменной не отвечает за то, для чего она нужна

dshestel replied:
нигде не используется, бесполезная переменная получается

dshestel commented on Assets/Scripts/BonusController.cs:
```diff
@@ -0,0 +1,23 @@
+    [SerializeField]
+    private BonusController _bonusController = null;
+    [SerializeField]
+    private int _bonusNumber = default;
```
тоже не совсем понятно для чего это переменная

dshestel commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -3,12 +3,20 @@
+
+    private void OnEnable()
+    {
+        BonusController.Spawed += BonusController_Spawed;
```
где отписка?

dshestel commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -33,6 +42,21 @@ private Vector3 GetMoveVector()
+          return UIJoystick.Instance.InputVector.x;
+    }
+
+    private void SpawnerPlayers(int CountPlayers)
```
название метода - непонятное
не по кодстайлу параметр входной

dshestel commented on Assets/Scripts/PoolController.cs:
```diff
@@ -0,0 +1,98 @@
+    public Queue<GameObject> Pool = new Queue<GameObject>();
+}
+
+public class PoolController : MonoBehaviour
```
будем переделывать

DenisHusainov committed:
Updated the names in the BonusController and PlayerController

dshestel commented on Assets/Scripts/BonusController.cs:
```diff
@@ -0,0 +1,29 @@
+    [SerializeField]
+    private BonusController _bonusController = null;
+    [SerializeField]
+    private int _setBonus = default;
```
setBonus - не понятное название

DenisHusainov committed:
Rewrote ObjectPool and create new Spawn in BonusController

dshestel commented on Assets/Scripts/BonusController.cs:
```diff
@@ -0,0 +1,28 @@
+
+    private void Start()
+    {
+        _bonusDoorText.text = _bonusSpawnNumber.ToString();
```
учим новую конструкцию. можно не править, но чтобы ты знал, что такое есть
_bonusDoorText.text = $"{_bonusDoorNumber}";

dshestel commented on Assets/Scripts/ObjectPool.cs:
```diff
@@ -0,0 +1,117 @@
+using UnityEngine;
+using System;
+
+public class ObjectPool<T> : IPool<T> where T : MonoBehaviour, IPoolable<T>
```
по код стайлу поправь

dshestel commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -1,14 +1,20 @@
+    private const float RightBorder = 3.5f;
+    private const float LeftBorder = -3.5f;
+
+    public static event Action CountSpawned = delegate { };
```
название ивента мне не нравится
количество заспаунилось
а количество чего - не понятно

dshestel commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -1,14 +1,20 @@
 
     private Vector3 _moveVector = default;
     private Vector3 _defaultSpeed = default;
+    private ObjectPool<PoolObject> _objectPool;
```
зачем тебе это тут?
у тебя эта штука должна лежать в самом пуле
и дергать из пула ты должен объекты, которые тебе нужны, которые кастятся к нужному типу и просто в лист складывать

dshestel commented on Assets/Scripts/PoolObject.cs:
```diff
@@ -0,0 +1,22 @@
+using System;
+using UnityEngine;
+
+public class PoolObject : MonoBehaviour, IPoolable<PoolObject>
```
название класса не понятное и по кодстайлу надо всё переписать

DenisHusainov committed:
Rename class, code style refactoring

dshestel commented on Assets/Scripts/ObjectPool.cs:
```diff
@@ -0,0 +1,117 @@
+{
+    private Action<T> _pullObject;
+    private Action<T> _pushObject;
+    private Stack<T> pooledObjects = new Stack<T>();
```
Dictionary<Poolable, Stack<Poolable>> ...

DenisHusainov committed:
Refactor PoolManager and fixed Singleton

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,69 @@
+
+    private void Spawn(int number)
+    {
+        PoolManager poolObject;
```
у тебя не будет вообще никаких пул манагеров больше

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,69 @@
+        Spawn(numToSpawn);
+    }
+
+    private void Spawn(int number)
```
этот метод должен иметь в качестве входных параментор объект, который надо заспаунить

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    public void Push(PoolManager pollObject)
```
он должен быть приватным

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,69 @@
+        stack.Push(pollObject);
+    }
+
+    public PoolManager Pull()
```
он должен быть дженериком, чтобы ты мог кастить сразу к нужному типу

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,69 @@
+
+        MonoBehaviour poolObject;
+
+        if (PooledCount > 0)
```
убирай каунт, потому что в это пуле будут лежать вообще все объекты, которые будут реализовывать интерфейч

DenisHusainov committed:
Fixed PoolManager, create new class for IPoolable

DenisHusainov committed:
Refactor Pull method

DenisHusainov committed:
Rewrote PoolManager with class Poolable

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,56 @@
+        Spawn(_prefab);
+    }
+
+    private void Spawn(Poolable objectForSpawn)
```
еще я посоветую вот тут еще передавать число
оно будет отвечать за то, сколько оъектов преперить
некоторые объекты надо преперить 50шт, а некоторых хватит и 5, будет не правильно хранить 50, где надо 5

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,56 @@
+
+    public T Pull<T>() where T : Poolable
+    {
+        if (_pooledObjects.TryGetValue(_prefab, out var stack))
```
сделай блок else
кидай ошибку(Debug.LogError), что такого не запреперено, преперь 1 штуку и возвращай
лучше не возвращать нал

DenisHusainov committed:
Fixed PoolManager: changed return function

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,67 @@
+    [SerializeField]
+    private Poolable _prefab = default;
+    [SerializeField]
+    private int _amountToPool = default;
```
это значение убираешь вообще, его не будет

dshestel commented on Assets/Scripts/PoolManager.cs:
```diff
@@ -0,0 +1,67 @@
+            Debug.LogError("This is not prepared");
+        }
+
+        var objectFromPool = (T)Instantiate(_prefab);
```
почему всю это логику ты не запихнул в блок else?

DenisHusainov committed:
Fixed PollManager

DenisHusainov committed:
Fixed PoolManager: Deleted field _amountToPool

DenisHusainov committed:
Fixed PoolManager: Renamed methods Pull and Push

DenisHusainov committed:
Create start/finish and add obstacle

DenisHusainov committed:
Created a class ObstacleController and sorted classes by folder

DenisHusainov committed:
Delete Log from commit

DenisHusainov committed:
Created a spawn in position

dshestel commented on Assets/Scripts/Managers/PoolManager.cs:
```diff
@@ -0,0 +1,82 @@
+
+    private void ObstaclesController_Crashed()
+    {
+        Prepare(_prefab);
```
нет. препейр проходит один раз на старте. ни ни какие ивенты это не проходит

dshestel commented on Assets/Scripts/ObstaclesController.cs:
```diff
@@ -0,0 +1,15 @@
+
+    private void OnTriggerEnter(Collider other)
+    {
+        if (other.gameObject.GetComponent<PlayerController>())
```
очень схожая логика с BonusController
объедени, потому что идет дублирование кода

dshestel commented on Assets/Scripts/Poolable.cs:
```diff
@@ -0,0 +1,19 @@
+using UnityEngine;
+
+public class Poolable : MonoBehaviour
```
где буль IsFree по которому ты определяешь активный ли объект на сцене или нет?

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+    private Rigidbody _rb = null;
+
+    private Vector3 _moveVector = default;
+    private Vector3 _defaultSpeed = default;
```
почему ты это сделал вектором?
когда нам надо знать только скорость по оси Z

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        _moveVector = GetMoveVector();
+        _rb.velocity = _defaultSpeed * Speed;
```
почему ты велосити сетишь в апдейте?
почему ты делаешь это через велосити?

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+
+        _moveVector = GetMoveVector();
+        _rb.velocity = _defaultSpeed * Speed;
+        transform.position = new Vector3(Mathf.Clamp(transform.position.x, LeftBorder, RightBorder), transform.position.y, transform.position.z);
```
а потом еще трансформ позишн сетишь

dshestel replied:
выбери одно, советую второе

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+        return dir;
+    }
+
+    public float HorizontalMove()
```
бесполезный публичный метод, всё это ты можешь дергать в методе выше

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+        return UIJoystick.Instance.InputVector.x;
+    }
+
+    public void Spawn(int count)
```
ISpawner.Spawn(...)
там будет более инкопсулированно

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    private bool CanMove()
```
почему ты это не можешь сделать пропертей?

DenisHusainov committed:
Optimized Glider Controller and combined bonuses and obstacles

DenisHusainov committed:
Refactor PlayerController: replaced with explicit implementation

dshestel commented on Assets/Scripts/Controllers/BonusAndObstalceController.cs:
```diff
@@ -0,0 +1,44 @@
+    {
+        if (other.gameObject.GetComponent<PlayerController>())
+        {
+            if (_type == TypesOfTrigger.Bonus)
```
давай тут наследрвание сделаем. 
сделай астрактный класс и от него наследуй, и не придётся какую-то логику через ифы делать

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,73 @@
+
+        _moveVector = GetMoveVector();
+        transform.position += _defaultSpeed * Speed;
+        transform.position = new Vector3(Mathf.Clamp(transform.position.x, LeftBorder, RightBorder), transform.position.y, transform.position.z);
```
ты два раза меняешь позицию. почему ты не можешь сначала просчитать expectedPosition, потом клэмпануть это, а только потом присвоить?

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,73 @@
+        transform.position += _moveVector;
+    }
+
+    private Vector3 GetMoveVector()
```
почему ты это делаешь через метод?
почему не можешь делать это в апдейте?
зачем ты возвращаешь Vector3 если там Vector2 и при всем этом меняется только по оси X, то есть там вообще должно возвращать просто флоат

dshestel commented on Assets/Scripts/Managers/PoolManager.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            Debug.LogError("This is not prepared");
+
+            objectFromPool = Instantiate(_prefab);
```
почему тут Instantiate?
сделай так, чтобы ты просто делал Spawn и тебе возвращался объект.
короче, один Instantiate на скрипт

DenisHusainov committed:
Fixed PlayerController: optimized Update

dshestel commented on Assets/Scripts/Controllers/Bonus.cs:
```diff
@@ -0,0 +1,6 @@
+using UnityEngine;
+
+public abstract class Bonus : MonoBehaviour
```
непонятное название класса, надо обдуманно классам давать нейинг

dshestel commented on Assets/Scripts/Controllers/PlayerController.cs:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        Vector3 expectedPosition = transform.position;
+        expectedPosition.z += DefaultSpeed;
```
я могу ошибаться, но персонаж по такой логике супер быстро передвигается, нету дельта тайма

DenisHusainov replied:
Вроде нормально он передвигается, там же DefaultSpeed=1f, получается на 1 с каждый кадром. Если дельтатайм добавлять, нужно умножать на скорость или DefaultSpeed делать больше. Просто если добавлять скорость получается лишняя переменная

dshestel replied:
тебе надо делать всё плавно, если я открою проект то будет все резко. надо плавнее делать

DenisHusainov committed:
Created abstract class Detected and divided the logic bonus and obstacle

DenisHusainov committed:
Fixed method Get() from PoolManager

dshestel commented on Assets/Scripts/Controllers/BonusController.cs:
```diff
@@ -0,0 +1,26 @@
+        _bonusDoorText.text = $"{_bonusSpawnNumber}";
+    }
+
+    public override void DetectedObject()
```
ты должен сюда пробрасывать реф на плеера, а там ты в онтригеринтер ты делаешь геткомпонент

dshestel commented on Assets/Scripts/GameWindow.cs:
```diff
@@ -17,5 +17,6 @@ private void Start()
     private void OnStartGameButtonClicked()
     {
         Started();
+        _startGameButton.gameObject.SetActive(false);
```
нет, ты должен открывать другое окно, а не скрывать компоненты на этом

dshestel commented on Assets/Scripts/Managers/PoolManager.cs:
```diff
@@ -0,0 +1,68 @@
+
+        if (_pooledObjects.TryGetValue(_prefab, out var stack))
+        {
+            var poolObject = (T)stack.Pop();
```
у тебя идет дублирование кода тут

DenisHusainov replied:
А где тут дублирование?

dshestel replied:
в блоке элс

dshestel commented on Assets/Scripts/Managers/PoolManager.cs:
```diff
@@ -0,0 +1,68 @@
+            Debug.LogError("This is not prepared");
+
+            Spawn(_prefab, 1);
+            objectFromPool = stack.Pop();
```
и тут, от этого надо избавиться

DenisHusainov committed:
Fixed BonusController

dshestel commented on Assets/Scripts/Controllers/Detected.cs:
```diff
@@ -0,0 +1,16 @@
+using UnityEngine;
+
+public abstract class Detected : MonoBehaviour
```
название класса как у ивента. продумывай названия классов

dshestel commented on Assets/Scripts/Controllers/Detected.cs:
```diff
@@ -0,0 +1,16 @@
+
+public abstract class Detected : MonoBehaviour
+{
+    public abstract void DetectedObject(PlayerController player);
```
название метода тоже не подходит. тут скорее подойдет OnInteracted(T t);

dshestel commented on Assets/Scripts/Controllers/Detected.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        var playerContoller = other.gameObject.GetComponent<PlayerController>();
+
+        if (playerContoller)
```
учимсчя использовать TryGetComponent

DenisHusainov committed:
Fixed PlayerController

added a smooth motion

DenisHusainov committed:
Renamed class and method

DenisHusainov committed:
Created MainWindow

Moved the start button to MainWindow and made a window transition

DenisHusainov commented on Assets/Scripts/Managers/UIManager.cs:
```diff
@@ -13,7 +23,21 @@ private void Awake()
 
     public void ShowWindow<T>() where T : Window
     {
+        CloseWindows();
```
Скорее всего это нужно по-другому делать, но так получилось. Если неправильно, отпиши как нужно, пожалуйста

dshestel replied:
грубо говоря всё так

DenisHusainov committed:
Fixed PoolManager

Fixed Get() method

dshestel commented on Assets/Scripts/Controllers/BonusController.cs:
```diff
@@ -0,0 +1,23 @@
+    [SerializeField]
+    private TextMeshPro _bonusDoorText = null;
+    [SerializeField]
+    private BonusController _bonusController = null;
```
зачем тебе ссылка на бонус контроллер?

DenisHusainov replied:
Он выключал второй бонус, который мы не используем, но лучше выключать сразу два

dshestel commented on Assets/Scripts/Controllers/BonusController.cs:
```diff
@@ -0,0 +1,23 @@
+    public override void OnInteracted(PlayerController player)
+    {
+        ((ISpawner)player).Spawn(_bonusSpawnNumber);
+        _bonusController.gameObject.SetActive(false);
```
почему нельзя просто сделать ГО.сет актив фалсе?

dshestel commented on Assets/Scripts/Controllers/Detected.cs:
```diff
@@ -0,0 +1,14 @@
+using UnityEngine;
+
+public abstract class ObjectContact : MonoBehaviour
```
название все еще не нравится

dshestel commented on Assets/Scripts/Controllers/ObstacleController.cs:
```diff
@@ -0,0 +1,11 @@
+using System;
+
+public class ObstacleController : ObjectContact
```
контроллер это сущность которая чем-то контролирует. тут у тебя подойдет порсто название обстакл

dshestel commented on Assets/Scripts/MainWindow.cs:
```diff
@@ -0,0 +1,22 @@
+
+    private void Start()
+    {
+        gameObject.SetActive(true);
```
у тебя эта логика должна быть на уровне окна, а не в старте. 
при последующих включениях экранов у тебя не будет работатт это

dshestel commented on Assets/Scripts/Managers/PoolManager.cs:
```diff
@@ -0,0 +1,65 @@
+
+    private void Start()
+    {
+        Prepare(_prefab);
```
у тебя по сути та же самая логика в эвэйке

DenisHusainov committed:
Fixed PoolManager and BonusController

Got rid of unnecessary logic in classes BonusController and optimized PoolManager

DenisHusainov committed:
Optimized GameManager

DenisHusainov committed:
Fixed UIManager

Made it so that at the beginning of the game all windows are closed and only the first one turns on

DenisHusainov committed:
Rename class
